### PR TITLE
rename minReplicationFactor to writeConcern

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Documentation companion PR: https://github.com/arangodb/docs/pull/181
Enterprise companion PR: https://github.com/arangodb/enterprise/pull/332

* Added startup option `--cluster.max-number-of-shards` for restricting the
  maximum number of shards when creating new collections. The default
  value for this setting is `1000`, which is also the previously hard-coded
  built-in limit. A value of `0` for this option means "unrestricted".
  When the setting is adjusted, it will not affect already existing
  collections, but only collections that are created or restored
  afterwards.

* Added startup options for managing the replication factor for newly 
  created collections:

  - `--cluster.min-replication-factor`: this settings controls the minimum
    replication factor value that is permitted when creating new collections.
    No collections can be created which have a replication factor value 
    below this setting's value. The default value is 1. 
  - `--cluster.max-replication-factor`: this settings controls the maximum
    replication factor value that is permitted when creating new collections.
    No collections can be created which have a replication factor value 
    above this setting's value. The default value is 10.
  - `--cluster.default-replication-factor`: this settings controls the default
    replication factor value that is used when creating new collections and 
    no value of replication factor has been specified. 
    If no value is set for this option, the value of the option
    `--cluster.min-replication-factor` will be used. 

Additionally renamed startup option `--cluster.min-replication-factor`
to `--cluster.write-concern` to make naming more consistent.

Internally, the write concern will still be persisted in an attribute
named "minReplicationFactor", because previous versions of ArangoDB
have already used that.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [ ] Bug-Fix for a *released version* (did you remember to port this to all relevant release branches?)
- [ ] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [x] Added new **integration tests** (shell_client, shell_server)

https://172.16.10.101/view/PR/job/arangodb-matrix-pr/6497/

### CODE REVIEW

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6565/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6568/

**Comment by jsteemann** [comment]
https://172.16.10.101/view/PR/job/arangodb-matrix-pr/6598/

**Comment by jsteemann** [comment]
Tests blue except unrelated http_server failure

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 1]
```diff
@@ -1170,35 +1178,42 @@ Result RestReplicationHandler::processRestoreCollectionCoordinator(
...
+  VPackSlice const replicationFactorSlice = parameters.get(StaticStrings::ReplicationFactor);
+  // not an error: for historical reasons the write concern is read from the
+  // variable "minReplicationFactor"
+  VPackSlice const writeConcernSlice = parameters.get(StaticStrings::MinReplicationFactor);
 
-  bool isValidReplFactorSlice = replFactorSlice.isInteger() ||
-                                (replFactorSlice.isString() &&
-                                 replFactorSlice.isEqualString(StaticStrings::Satellite));
+  bool isValidReplicationFactorSlice = replicationFactorSlice.isInteger() ||
+                                       (replicationFactorSlice.isString() &&
```
Comment (mchacki) [inline arangod/RestHandler/RestReplicationHandler.cpp]: this is easier to understand :+1:

File: arangod/Sharding/ShardingInfo.cpp [hunk 2]
```diff
@@ -128,25 +139,27 @@ ShardingInfo::ShardingInfo(arangodb::velocypack::Slice info, LogicalCollection*
...
-      if (!isSatellite() && _minReplicationFactor > _replicationFactor) {
+  if (!isASatellite) {
+    auto writeConcernSlice = info.get(StaticStrings::MinReplicationFactor);
+    if (!writeConcernSlice.isNone()) {
+      if (writeConcernSlice.isNumber()) {
+        _writeConcern = writeConcernSlice.getNumber<size_t>();
+        if (!isSatellite() && _writeConcern > _replicationFactor) {
+          THROW_ARANGO_EXCEPTION_MESSAGE(
+              TRI_ERROR_BAD_PARAMETER,
+              "writeConcern cannot be larger than replicationFactor (" +
```
Comment (mchacki) [inline arangod/Sharding/ShardingInfo.cpp]: Do we want to name it writeConcern here?
as far as I see this is only the "internal" name and we have to stick to minRepicationFactor over API, so I guess user might be confused?

File: arangod/Sharding/ShardingInfo.cpp [hunk 3]
```diff
@@ -128,25 +139,27 @@ ShardingInfo::ShardingInfo(arangodb::velocypack::Slice info, LogicalCollection*
...
+          THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_BAD_PARAMETER,
+                                         "writeConcern cannot be 0");
+        }
+      } else {
         THROW_ARANGO_EXCEPTION_MESSAGE(
             TRI_ERROR_BAD_PARAMETER,
-            "minReplicationFactor cannot be larger then replicationFactor (" +
-                basics::StringUtils::itoa(_minReplicationFactor) + " > " +
-                basics::StringUtils::itoa(_replicationFactor) + ")");
+            "writeConcern needs to be an integer number");
```
Comment (mchacki) [inline arangod/Sharding/ShardingInfo.cpp]: Same here, we want to name it writeConcern or stick to minReplicationFactor?

File: arangod/Sharding/ShardingInfo.cpp [hunk 4]
```diff
@@ -360,47 +373,47 @@ void ShardingInfo::avoidServers(std::vector<std::string> const& avoidServers) {
...
   return _replicationFactor;
 }
 
 void ShardingInfo::replicationFactor(size_t replicationFactor) {
-  if (!isSatellite() && replicationFactor < _minReplicationFactor) {
+  if (!isSatellite() && replicationFactor < _writeConcern) {
     THROW_ARANGO_EXCEPTION_MESSAGE(
         TRI_ERROR_BAD_PARAMETER,
-        "replicationFactor cannot be smaller then minReplicationFactor (" +
+        "replicationFactor cannot be smaller than writeConcern (" +
```
Comment (mchacki) [inline arangod/Sharding/ShardingInfo.cpp]: again here.

File: arangod/Sharding/ShardingInfo.cpp [hunk 5]
```diff
@@ -360,47 +373,47 @@ void ShardingInfo::avoidServers(std::vector<std::string> const& avoidServers) {
...
 
-void ShardingInfo::minReplicationFactor(size_t minReplicationFactor) {
-  if (!isSatellite() && minReplicationFactor > _replicationFactor) {
+void ShardingInfo::writeConcern(size_t writeConcern) {
+  if (!isSatellite() && writeConcern > _replicationFactor) {
     THROW_ARANGO_EXCEPTION_MESSAGE(
         TRI_ERROR_BAD_PARAMETER,
-        "minReplicationFactor cannot be larger then replicationFactor (" +
-            basics::StringUtils::itoa(_minReplicationFactor) + " > " +
+        "writeConcern cannot be larger than replicationFactor (" +
```
Comment (mchacki) [inline arangod/Sharding/ShardingInfo.cpp]: again

File: arangod/Sharding/ShardingInfo.cpp [hunk 6]
```diff
@@ -360,47 +373,47 @@ void ShardingInfo::avoidServers(std::vector<std::string> const& avoidServers) {
...
-void ShardingInfo::setMinAndMaxReplicationFactor(size_t minimal, size_t maximal) {
-  if (minimal > maximal) {
+void ShardingInfo::setWriteConcernAndReplicationFactor(size_t writeConcern, size_t replicationFactor) {
+  if (writeConcern > replicationFactor) {
     THROW_ARANGO_EXCEPTION_MESSAGE(
         TRI_ERROR_BAD_PARAMETER,
-        "minReplicationFactor cannot be larger then replicationFactor (" +
-            basics::StringUtils::itoa(minimal) + " > " +
-            basics::StringUtils::itoa(maximal) + ")");
+        "writeConcern cannot be larger than replicationFactor (" +
```
Comment (mchacki) [inline arangod/Sharding/ShardingInfo.cpp]: again

File: arangod/Transaction/Methods.cpp [hunk 7]
```diff
@@ -1604,9 +1604,9 @@ Future<OperationResult> transaction::Methods::insertLocal(std::string const& cna
         // We cannot fulfill minimum replication Factor.
         // Reject write.
         LOG_TOPIC("d7306", ERR, Logger::REPLICATION)
-            << "Less then minReplicationFactor "
-            << basics::StringUtils::itoa(collection->minReplicationFactor())
-            << " followers in sync. Shard  " << collection->name()
+            << "Less than writeConcern ("
```
Comment (mchacki) [inline arangod/Transaction/Methods.cpp]: All logging in this file, do we want to publicly change it to writeConcern?

File: arangod/VocBase/LogicalCollection.cpp [hunk 8]
```diff
@@ -798,43 +800,42 @@ arangodb::Result LogicalCollection::properties(velocypack::Slice const& slice,
...
-      if (minrfTest < 0) {
-        // negative value for min replication factor... not good
+  if (!wcSl.isNone()) {
+    if (wcSl.isInteger()) {
+      int64_t wcTest = wcSl.getNumber<int64_t>();
+      if (wcTest < 0) {
+        // negative value for writeConcern... not good
         return Result(TRI_ERROR_BAD_PARAMETER,
-                      "bad value for minReplicationFactor");
+                      "bad value for writeConcern");
```
Comment (mchacki) [inline arangod/VocBase/LogicalCollection.cpp]: again, visible change?

File: arangod/VocBase/LogicalCollection.cpp [hunk 9]
```diff
@@ -798,43 +800,42 @@ arangodb::Result LogicalCollection::properties(velocypack::Slice const& slice,
...
       }
 
-      minrf = minrfSl.getNumber<size_t>();
-      if (minrf > rf || minrf > 10) {
+      wc = wcSl.getNumber<size_t>();
+      if (wc > rf) {
         return Result(TRI_ERROR_BAD_PARAMETER,
-                      "bad value for minReplicationFactor");
+                      "bad value for writeConcern");
```
Comment (mchacki) [inline arangod/VocBase/LogicalCollection.cpp]: again, visible change?

File: arangod/VocBase/LogicalCollection.cpp [hunk 10]
```diff
@@ -798,43 +800,42 @@ arangodb::Result LogicalCollection::properties(velocypack::Slice const& slice,
...
       }
 
       if (ServerState::instance()->isCoordinator() &&
-          rf != _sharding->minReplicationFactor()) {  // sanity checks
+          rf != _sharding->writeConcern()) {  // sanity checks
         if (!_sharding->distributeShardsLike().empty()) {
           return Result(TRI_ERROR_FORBIDDEN,
-                        "Cannot change minReplicationFactor, "
-                        "please change " +
+                        "Cannot change writeConcern, please change " +
```
Comment (mchacki) [inline arangod/VocBase/LogicalCollection.cpp]: again, visible change?

File: arangod/VocBase/LogicalCollection.cpp [hunk 11]
```diff
@@ -798,43 +800,42 @@ arangodb::Result LogicalCollection::properties(velocypack::Slice const& slice,
...
                             _sharding->distributeShardsLike());
         } else if (_type == TRI_COL_TYPE_EDGE && _isSmart) {
           return Result(TRI_ERROR_NOT_IMPLEMENTED,
-                        "Changing minReplicationFactor "
+                        "Changing writeConcern "
```
Comment (mchacki) [inline arangod/VocBase/LogicalCollection.cpp]: again, visible change?

File: arangod/VocBase/LogicalCollection.cpp [hunk 12]
```diff
@@ -798,43 +800,42 @@ arangodb::Result LogicalCollection::properties(velocypack::Slice const& slice,
...
                         "not supported for smart edge collections");
         } else if (isSatellite()) {
           return Result(TRI_ERROR_FORBIDDEN,
                         "Satellite collection, "
-                        "cannot change minReplicationFactor");
+                        "cannot change writeConcern");
```
Comment (mchacki) [inline arangod/VocBase/LogicalCollection.cpp]: again, visible change?

File: arangod/VocBase/LogicalCollection.cpp [hunk 13]
```diff
@@ -798,43 +800,42 @@ arangodb::Result LogicalCollection::properties(velocypack::Slice const& slice,
...
         }
       }
     } else {
       return Result(TRI_ERROR_BAD_PARAMETER,
-                    "bad value for minReplicationFactor");
+                    "bad value for writeConcern");
```
Comment (mchacki) [inline arangod/VocBase/LogicalCollection.cpp]: again, visible change?

File: arangod/VocBase/VocbaseInfo.cpp [hunk 14]
```diff
@@ -343,19 +343,21 @@ VocbaseOptions getVocbaseOptions(application_features::ApplicationServer& server
...
-        (minReplicationSlice.isNumber() && minReplicationSlice.getNumber<int>() > 0);
+        (writeConcernSlice.isNumber() && writeConcernSlice.getNumber<int>() > 0);
     if (!isNumber) {
       if (haveCluster) {
-        vocbaseOptions.minReplicationFactor = server.getFeature<ClusterFeature>().minReplicationFactor();
+        vocbaseOptions.writeConcern = server.getFeature<ClusterFeature>().writeConcern();
       } else {
         LOG_TOPIC("eeeed", ERR, Logger::CLUSTER)
-            << "Cannot access ClusterFeature to determine minReplicationFactor";
+            << "Cannot access ClusterFeature to determine writeConcern";
```
Comment (mchacki) [inline arangod/VocBase/VocbaseInfo.cpp]: again, visible change?

File: arangod/VocBase/vocbase.cpp [hunk 15]
```diff
@@ -1812,13 +1812,13 @@ arangodb::Result TRI_vocbase_t::toVelocyPack(VPackBuilder& result) const {
   {
     VPackObjectBuilder b(&result);
 
-    result.add("name", VPackValue(_name));
-    result.add("id", VPackValue(std::to_string(_id)));
-    result.add("isSystem", VPackValue(isSystem()));
+    result.add(StaticStrings::DataSourceName, VPackValue(_name));
+    result.add(StaticStrings::DataSourceId, VPackValue(std::to_string(_id)));
+    result.add(StaticStrings::DataSourceSystem, VPackValue(isSystem()));
```
Comment (mchacki) [inline arangod/VocBase/vocbase.cpp]: Yay \o/

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/collectionsView.js [hunk 16]
```diff
@@ -430,9 +443,9 @@
 
           var abort = false;
           try {
-            if (Number.parseInt(minReplicationFactor) > Number.parseInt(replicationFactor)) {
+            if (Number.parseInt(writeConcern) > Number.parseInt(replicationFactor)) {
               // validation here, as our Joi integration misses some core features
-              arangoHelper.arangoError("New Collection", "Minimal replication factor is not allowed to be greater than replication factor");
+              arangoHelper.arangoError("New Collection", "Write concern is not allowed to be greater than replication factor");
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/collectionsView.js]: again, visible change?

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/collectionsView.js [hunk 17]
```diff
@@ -625,8 +645,8 @@
 
             advancedTableContent.push(
               window.modalView.createTextEntry(
-                'new-min-replication-factor',
-                'Mininum replication factor',
+                'new-write-concern',
+                'Write concern',
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/collectionsView.js]: again, visible change?

So in this way our API names and the WebUI names differ.
I do not really like this...

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/databaseView.js [hunk 18]
```diff
@@ -407,19 +404,17 @@
 
         tableContent.push(
           window.modalView.createTextEntry(
-            'new-min-replication-factor',
-            'Mininum replication factor',
+            'new-write-concern',
+            'Write concern',
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/databaseView.js]: again, visible change?
Also here now the WebUI shows different name than API

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/databaseView.js [hunk 19]
```diff
@@ -407,19 +404,17 @@
...
             dbDefaultProperties.minReplicationFactor,
             'Numeric value. Must be at least 1 and must be smaller or equal compared to the replicationFactor. Minimal number of copies of the data in the cluster to be in sync in order to allow writes.',
-            'Default minimum replication factor',
+            'Default write concern',
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/databaseView.js]: again, visible change?

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/graphManagementView.js [hunk 20]
```diff
@@ -894,8 +894,8 @@
 
         tableContent.push(
           window.modalView.createTextEntry(
-            'new-minReplicationFactor',
-            'Minimal Replication factor',
+            'new-writeConcern',
+            'Write concern',
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/graphManagementView.js]: again, visible change?

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/graphManagementView.js [hunk 21]
```diff
@@ -821,8 +821,8 @@
         if (graph.get('minReplicationFactor')) {
           tableContent.push(
             window.modalView.createReadOnlyEntry(
-              'minReplicationFactor',
-              'Minimal Replication factor',
+              'writeConcern',
+              'Write concern',
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/graphManagementView.js]: again, visible change?

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/graphManagementView.js [hunk 22]
```diff
@@ -964,8 +964,8 @@
         );
         tableContent.push(
           window.modalView.createTextEntry(
-            'general-minReplicationFactor',
-            'Minimal replication factor',
+            'general-writeConcern',
+            'Write concern',
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/graphManagementView.js]: again, visible change?

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/settingsView.js [hunk 23]
```diff
@@ -163,22 +163,22 @@
...
                   if (frontendConfig.isCluster) {
                     replicationFactor = $('#change-replication-factor').val();
-                    minReplicationFactor = $('#change-min-replication-factor').val();
+                    writeConcern = $('#change-write-concern').val();
                     try {
-                      if (Number.parseInt(minReplicationFactor) > Number.parseInt(replicationFactor)) {
+                      if (Number.parseInt(writeConcern) > Number.parseInt(replicationFactor)) {
                         // validation here, as our Joi integration misses some core features
-                        arangoHelper.arangoError("Change Collection", "Minimal replication factor is not allowed to be greater then replication factor");
+                        arangoHelper.arangoError("Change Collection", "Write concern is not allowed to be greater then replication factor");
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/settingsView.js]: again, visible change?

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/settingsView.js [hunk 24]
```diff
@@ -428,10 +428,10 @@
                     );
                     tableContent.push(
                       window.modalView.createReadOnlyEntry(
-                        'change-min-replication-factor',
-                        'Minimal replication factor',
-                        data.minReplicationFactor,
-                        'This collection is a satellite collection. The minReplicationFactor is not changeable.',
+                        'change-write-concern',
+                        'Write concern',
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/settingsView.js]: again, visible change?

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/settingsView.js [hunk 25]
```diff
@@ -428,10 +428,10 @@
...
+                        data.writeConcern,
+                        'This collection is a satellite collection. The writeConcern is not changeable.',
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/settingsView.js]: again, visible change?

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/settingsView.js [hunk 26]
```diff
@@ -455,10 +455,10 @@
                     );
                     tableContent.push(
                       window.modalView.createTextEntry(
-                        'change-min-replication-factor',
-                        'Minimal Replication factor',
-                        data.minReplicationFactor,
-                        'Numeric value. Must be at least 1. Must be smaller or equal compared to the replicationFactor. Total number of copies of the data in the cluster. If we get below this value the collection will be read-only until enough copies are created.',
+                        'change-write-concern',
+                        'Write concern',
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/settingsView.js]: again, visible change?

File: tests/js/common/shell/shell-cluster-collection.js [hunk 27]
```diff
@@ -493,29 +494,83 @@ function ClusterCollectionSuite () {
...
-////////////////////////////////////////////////////////////////////////////////
-/// @brief test create
-////////////////////////////////////////////////////////////////////////////////
-
     testCreateInvalidNumberOfShards1 : function () {
       try {
         db._create("UnitTestsClusterCrud", { numberOfShards : 0 });
-      }
-      catch (err) {
+        fail();
```
Comment (mchacki) [inline tests/js/common/shell/shell-cluster-collection.js]: great these tests have been pointless before ...

**Comment by jsteemann** [comment]
https://172.16.10.101/view/PR/job/arangodb-matrix-pr/6610/

**Pull request merged**